### ISSUE TITLE
Create a nested directory while saving objects using `torch.save`

### ISSUE THREAD

**Issue by take2rohit** [pytorch/pytorch#71389]
### 🚀 The feature, motivation and pitch

Currently, when we try to save anything using `torch.save` method for which the full path doesn't exist, it throws an error. 
We then need to first create the directory using methods like `os.makedirs` and then use `torch.save`. This makes the code seems less elegant. 

Since `torch.save` is a heavily used method,  I feel this small addition will be very valuable as it will reduce the amount of code and enhance readability. Let me know what you guys think about this. If you feel it's worth adding to the library let me know, I will create a PR for the same.

### Alternatives

Let's take an example

The current saving code will be 
```python
import torch
import os

tensor = torch.zeros(10,10)
path = 'weights/DomainAdaptation/Office31/amazon2webcam'
os.makedirs(path, exist_ok = True) 
torch.save(tensor , os.path.join(path,'model.pth'))
```

I want this to be replaced by a single line that takes care of creating paths if the path doesn't exist. This can be done by adding an additional argument say `create_dir`. So the above code should become 

```python
import torch
tensor = torch.zeros(10,10)
torch.save(tensor,'weights/DomainAdaptation/Office31/amazon2webcam/model.pth', create_dir=True)
```

**Comment by mruberry** [comment]
This makes sense and I appreciate why this little refinement would be neat, but we try to keep PyTorch simple and promote one best way to accomplish a task. The pattern of not saving to a directory unless it already exists is pretty standard, too.

Neat idea, though @take2rohit.

**Comment by SKPsanjeevi** [comment]
@mruberry 
1. I just noticed this to be standard across Python. i.e. sub-folders cannot be created automatically. May I know why this is not enabled (or doable) by default?
2. If this is not going to be implemented, will this issue need to be closed?
PS: Sorry I am new here. Trying to understand the process flow.

**Comment by swissarthurfreeman** [comment]
+1
It is rather inconvenient to have to run os.makedir everytime, this will result in thousands of redundant lines of code in projects.

**Comment by ash-01xor** [comment]
Yes would make things easier while working with models, i  sometimes wish to not create a directory each time before hand.

If there is no one willing to work on this issue , i am willing to work on this , hope i do get guidance while working on it.

**Comment by ph-code-repo** [comment]
even if this functionality was added and toggled by a boolean function argument, this would be really handy for my work at the moment too.

**Comment by casszhao** [comment]
> +1 It is rather inconvenient to have to run os.makedir everytime, this will result in thousands of redundant lines of code in projects.

+1

**Comment by OliEfr** [comment]
+1

**Comment by yyu233** [comment]
+1

**Comment by DutraIRS** [comment]
+1

**Comment by jeffreygwang** [comment]
+1

**Comment by annlaurin** [comment]
+1

**Comment by oskar-codes** [comment]
+1
